twilight-gateway/src/future.rs

```diff
@@ -0,0 +1,206 @@
+
+            return Poll::Ready(NextMessageFutureOutput::Message(maybe_message));
+        }
+
+        Poll::Pending
+    }
+}
+
+/// Future that will resolve when the delay for a reconnect passes.
+///
+/// The duration of the future is defined by the number of attempts at
+/// reconnecting that have already been made. The math behind it is
+/// `2 ^ attempts`, maxing out at [`MAX_WAIT_SECONDS`].
+///
+/// [`MAX_WAIT_SECONDS`]: Self::MAX_WAIT_SECONDS
+pub struct ReconnectDelayFuture {
+    /// Inner future resolving when the duration passes.
+    inner: Pin<Box<Sleep>>,
+}
+
+impl ReconnectDelayFuture {
+    /// The maximum wait before resolving, in seconds.
+    const MAX_WAIT_SECONDS: u8 = 128;
+
+    /// Initialize a new unpolled future that will resolve when a reconnect
+    /// should be made.
+    pub fn new(reconnect_attempts: u8) -> Self {
+        let wait = 2_u8
+            .saturating_pow(u32::from(reconnect_attempts))
+            .min(Self::MAX_WAIT_SECONDS);
+
+        let duration = Duration::from_secs(u64::from(wait));
+
+        Self {
+            inner: Box::pin(time::sleep(duration)),
+        }
+    }
+}
+
+impl Future for ReconnectDelayFuture {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.as_mut().inner.poll_unpin(cx)
+    }
+}
+
+/// Future that will resolve when the shard must send its next heartbeat.
+///
+/// The duration of the future is defined by taking the heartbeat interval defined
+/// by the [`Ready`] event, and subtracting the duration since the
+/// [last heartbeat] was sent. If a [`Ready`] event has not yet been received
+/// then the future will never be ready, and if the duration has already passed
+/// then it will immediately be ready.
+///
+/// This future must always take precedence over other actions in order to
+/// maintain the [gateway session]!
+///
+/// [gateway session]: crate::Session
+/// [`Ready`]: twilight_model::gateway::payload::incoming::Ready
+pub struct TickHeartbeatFuture {
+    /// Inner future that will resolve after some time, defined by the type-level
+    /// documentation.
+    inner: Option<Pin<Box<Sleep>>>,
+}
+
+impl TickHeartbeatFuture {
+    /// Initialize a new unpolled future that will resolve when the next
+    /// heartbeat must be sent.
+    pub fn new(
+        maybe_last_sent: Option<Instant>,
+        maybe_heartbeat_interval: Option<Duration>,
+    ) -> Self {
+        let heartbeat_interval = if let Some(heartbeat_interval) = maybe_heartbeat_interval {
+            heartbeat_interval
+        } else {
+            return Self { inner: None };
+        };
+
+        let remaining = if let Some(last_sent) = maybe_last_sent {
+            let time_since = last_sent.elapsed();
+
+            heartbeat_interval.saturating_sub(time_since)
+        } else {
+            Duration::ZERO
+        };
+
+        Self {
+            inner: Some(Box::pin(time::sleep(remaining))),
```

Comment:
This allocation can be removed if we use `pin-project-lite` (transiently depended on through tokio), note however that it currently does not support adding comments to fields.
I've already made a patch that does this which I can send.